File: apps/ios/Sources/BikeCore/WorkspaceFactory.swift
import Foundation

public func outlineNode(
    _ text: String,
    note: String = "",
    children: [OutlineNode] = []
) -> OutlineNode {
    OutlineNode(text: text, note: note, children: children)
}

public func createStarterWorkspace(now: Date = Date()) -> Workspace {
    let timestamp = ISO8601DateFormatter.bike.string(from: now)
    let documentId = newBikeId("doc")
    let document = OutlineDocument(
        id: documentId,
        title: inboxDocumentTitle,
        createdAt: timestamp,
        updatedAt: timestamp,
        nodes: [
            outlineNode(
                "快速捕捉",
                children: [
                    outlineNode("从 iOS 分享面板收集文字和链接"),
                    outlineNode("离线保存到本机工作区"),
                    outlineNode("回到桌面端继续深度整理")
                ]
            ),
            outlineNode(
                "轻量整理",
                children: [
                    outlineNode("新增同级和子级主题"),
                    outlineNode("折叠、勾选、备注"),
                    outlineNode("保持 Workspace v1 JSON 兼容")
                ]
            )
        ]
    )
    return Workspace(activeDocumentId: documentId, documents: [document])
}

public extension Workspace {
    func activeDocument() throws -> OutlineDocument {
        if let document = documents.first(where: { $0.id == activeDocumentId }) {
            return document
        }
        if let document = documents.first {
            return document
        }
        throw WorkspaceError.noAvailableDocument
    }
}

public extension OutlineDocument {
    func nodeCount() -> Int {
        nodes.reduce(0) { $0 + $1.nodeCount() }
    }
}

public extension OutlineNode {
    func nodeCount() -> Int {
        1 + children.reduce(0) { $0 + $1.nodeCount() }
    }
}

public enum WorkspaceError: LocalizedError, Equatable {
    case noAvailableDocument
    case invalidWorkspace

    public var errorDescription: String? {
        switch self {
        case .noAvailableDocument:
            "工作区没有可用文档"
        case .invalidWorkspace:
            "工作区至少需要一篇文档"
        }
    }
}

File: apps/android/app/src/main/java/com/bike/android/data/WorkspaceFactory.kt
package com.bike.android.data

import java.time.Instant
import java.util.UUID

const val INBOX_DOCUMENT_TITLE = "Bike Android 收件箱"

fun createStarterWorkspace(now: Instant = Instant.now()): Workspace {
    val timestamp = now.toString()
    val documentId = newBikeId("doc")

    return Workspace(
        activeDocumentId = documentId,
        documents = listOf(
            OutlineDocument(
                id = documentId,
                title = INBOX_DOCUMENT_TITLE,
                createdAt = timestamp,
                updatedAt = timestamp,
                nodes = listOf(
                    outlineNode(
                        text = "快速捕捉",
                        children = listOf(
                            outlineNode("从 Android 分享面板收集文字和链接"),
                            outlineNode("离线保存到本机工作区"),
                            outlineNode("回到桌面端继续深度整理"),
                        ),
                    ),
                    outlineNode(
                        text = "轻量整理",
                        children = listOf(
                            outlineNode("新增同级和子级主题"),
                            outlineNode("折叠、勾选、备注"),
                            outlineNode("保持 Workspace v1 JSON 兼容"),
                        ),
                    ),
                ),
            ),
        ),
    )
}

fun outlineNode(
    text: String,
    children: List<OutlineNode> = emptyList(),
): OutlineNode =
    OutlineNode(
        id = newBikeId("node"),
        text = text,
        children = children,
    )

fun outlineNode(
    text: String,
    note: String,
    children: List<OutlineNode> = emptyList(),
): OutlineNode =
    outlineNode(text = text, children = children).copy(note = note)

fun Workspace.activeDocument(): OutlineDocument =
    documents.firstOrNull { it.id == activeDocumentId }
        ?: documents.firstOrNull()
        ?: error("工作区没有可用文档")

fun OutlineDocument.nodeCount(): Int =
    nodes.sumOf { it.nodeCount() }

fun OutlineNode.nodeCount(): Int =
    1 + children.sumOf { it.nodeCount() }

internal fun newBikeId(prefix: String): String =
    "${prefix}_${UUID.randomUUID()}"
